### DISCUSSION TITLE
Renaming gamepad button constants?

### DISCUSSION THREAD

**Discussion by SolarLune** [hajimehoshi/ebiten#1943] [Ideas]
Hello!

Would it be possible to rename the standard gamepad constants to be something a bit more understandable? "StandardGamepadButtonRightBottom" is perfectly functional, but "StandardGamepadA" or "StandardGamepadCross" is easier to immediately understand. While it would, of course, be different depending on what gamepad you're using, I think it's easier to mentally translate controller constants that refer to an Xbox layout to a PS layout (or vice-versa) over translating what "StandardGamepadButtonLeftTop" is referring to.  Using "StandardGamepadFront*" to refer to the shoulder button area is also a bit weird to wrap my mind around.

**Comment by hajimehoshi** [comment]
The button name is based on [the Web standard](https://www.w3.org/TR/gamepad/#remapping). This might be weird but this should be universal and stable. I don't plan to change the names so far. If you really want to have a better name, you can define constants by yourself.

> StandardGamepadA

This has different meanings between Nintendo Switch and Xbox, right?

**Reply by SolarLune** [reply]
> The button name is based on the Web standard. This might be weird but this should be universal and stable. I don't plan to change the names so far. If you really want to have a better name, you can define constants by yourself.

Fair enough, understandable. EDIT: Actually, the graphic on that page helps to visualize the button locations - would it be possible to add this to the documentation or codebase as a link in a comment?

> This has different meanings between Nintendo Switch and Xbox, right?

Yeah, just like "X" (or cross) on a PS5 controller and "X" on a Switch and Xbox controller represents different locations on their respective controllers. I just think it would be easier to conceptualize "the A button on Xbox, but on whatever controller Ebiten calls the default (usually Xbox)" rather than a series of words indicating the buttons' locations on the controller.

**Reply by hajimehoshi** [reply]
> Fair enough, understandable. EDIT: Actually, the graphic on that page helps to visualize the button locations - would it be possible to add this to the documentation or codebase as a link in a comment?

It's already linked: https://pkg.go.dev/github.com/hajimehoshi/ebiten/v2#StandardGamepadButton

**Reply by hajimehoshi** [reply]
>  I just think it would be easier to conceptualize "the A button on Xbox, but on whatever controller Ebiten calls the default (usually Xbox)" rather than a series of words indicating the buttons' locations on the controller.

If there was not the standard, I'd adopt Xbox as a reference.

**Reply by SolarLune** [reply]
> It's already linked: https://pkg.go.dev/github.com/hajimehoshi/ebiten/v2#StandardGamepadButton

Fantastic, that's very helpful; I didn't think to check the comments, my bad. Thanks!

**Comment by eliasdaler** [comment]
Maybe others will find it useful. I made the following constants/mapping to make it easier for myself:

```go
type GamepadButton int

const (
	GamepadButtonDPadUp = iota
	GamepadButtonDPadDown
	GamepadButtonDPadLeft
	GamepadButtonDPadRight
	GamepadButtonTop           // Y (Xbox), triangle (PS)
	GamepadButtonBottom        // A (Xbox), cross (PS)
	GamepadButtonLeft          // X (Xbox), square (PS)
	GamepadButtonRight         // B (Xbox), circle (PS)
	GamepadButtonLeftShoulder  // L1
	GamepadButtonRightShoulder // R1
	GamepadButtonLeftTrigger   // L2
	GamepadButtonRightTrigger  // R2
	GamepadButtonSelect        // Share on PS
	GamepadButtonStart         // Options on PS
	GamepadButtonLeftStick     // L3 on PS
	GamepadButtonRightStick    // R3 on PS
	GamepadButtonHome
)

var gamepadButtonToEbitenStandardButton = map[GamepadButton]ebiten.StandardGamepadButton{
	GamepadButtonDPadUp:        ebiten.StandardGamepadButtonLeftTop,
	GamepadButtonDPadDown:      ebiten.StandardGamepadButtonLeftBottom,
	GamepadButtonDPadLeft:      ebiten.StandardGamepadButtonLeftLeft,
	GamepadButtonDPadRight:     ebiten.StandardGamepadButtonLeftRight,
	GamepadButtonTop:           ebiten.StandardGamepadButtonRightTop,
	GamepadButtonBottom:        ebiten.StandardGamepadButtonRightBottom,
	GamepadButtonLeft:          ebiten.StandardGamepadButtonRightLeft,
	GamepadButtonRight:         ebiten.StandardGamepadButtonRightRight,
	GamepadButtonLeftShoulder:  ebiten.StandardGamepadButtonFrontTopLeft,
	GamepadButtonRightShoulder: ebiten.StandardGamepadButtonFrontTopRight,
	GamepadButtonLeftTrigger:   ebiten.StandardGamepadButtonFrontBottomLeft,
	GamepadButtonRightTrigger:  ebiten.StandardGamepadButtonFrontBottomRight,
	GamepadButtonSelect:        ebiten.StandardGamepadButtonCenterLeft,
	GamepadButtonStart:         ebiten.StandardGamepadButtonCenterRight,
	GamepadButtonLeftStick:     ebiten.StandardGamepadButtonLeftStick,
	GamepadButtonRightStick:    ebiten.StandardGamepadButtonRightStick,
	GamepadButtonHome:          ebiten.StandardGamepadButtonCenterCenter,
}
```
